src/Aspire.Hosting/Dashboard/KubernetesDataSource.cs

```diff
@@ -0,0 +1,593 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Concurrent;
+using System.Diagnostics;
+using System.Text;
+using System.Text.Json;
+using System.Text.Json.Nodes;
+using Aspire.Dashboard.Model;
+using Aspire.Hosting.ApplicationModel;
+using Aspire.Hosting.Dcp;
+using Aspire.Hosting.Dcp.Model;
+using Aspire.Hosting.Dcp.Process;
+using Aspire.Hosting.Utils;
+using k8s;
+using Microsoft.Extensions.Logging;
+
+namespace Aspire.Hosting.Dashboard;
+
+/// <summary>
+/// Pulls data about kubernetes resources from DCP's kubernetes API. Streams updates to consumers.
+/// </summary>
+internal sealed class KubernetesDataSource
+{
+    private readonly KubernetesService _kubernetesService;
+    private readonly DistributedApplicationModel _applicationModel;
+    private readonly Func<ResourceViewModel, ObjectChangeType, ValueTask> _onResourceChanged;
+    private readonly CancellationToken _cancellationToken;
+    private readonly ILogger _logger;
+
+    private readonly Dictionary<string, Container> _containersMap = [];
+    private readonly Dictionary<string, Executable> _executablesMap = [];
+    private readonly Dictionary<string, Service> _servicesMap = [];
```

Comment:
I'm wondering how pervasive we want to change the model from inheritance based to composition. Two things are pushing us in the composition direction:
* Generic resources
* gRPC doesn't support inheritance

At what layer does the idea of a generic resource begin?

(note: just posing the question in this PR. I don't think change needs to happen right now)